### ISSUE TITLE
[gfx950][Kimi-K3] One-pass Gluon MLA prefill: 50K/500 benchmark results

### ISSUE THREAD

**Issue by raikonenfnu** [raikonenfnu/tokenspeed#88]
## Summary

The one-pass full-history MLA prefill path in [`e3185b83`](https://github.com/raikonenfnu/tokenspeed/commit/e3185b838cbfdbce873aa07498325414ea939264) improves end-to-end output throughput by **21.8%–44.4%** across concurrency 1/2/4/8/16 relative to clean main at `cae0f292`.

This applies the runtime-side optimization explored in #87 while using TokenSpeed's existing native gfx950 Gluon MLA and KDA kernels—no AITER dependency or port.

## Output throughput: before vs. after

![Grouped bar chart comparing Kimi-K3 output throughput before and after one-pass Gluon MLA prefill](https://raw.githubusercontent.com/raikonenfnu/tokenspeed/04e18f2e/docs/assets/kimi-k3-prefill-before-after.svg)

Values are aggregate output tokens/s across the TP8 server; higher is better.

## TTFT, TPOT, and total latency: before vs. after

![Three grouped bar charts comparing Kimi-K3 TTFT, TPOT, and total latency before and after one-pass Gluon MLA prefill](https://raw.githubusercontent.com/raikonenfnu/tokenspeed/9251ada0/docs/assets/kimi-k3-prefill-latency-before-after.svg)

Each panel has an independent scale and labeled units; lower is better.

| Concurrency | Baseline tok/s | Optimized tok/s | Throughput gain | TTFT change | TPOT change | Avg latency change |
|---:|---:|---:|---:|---:|---:|---:|
| 1 | 19.17 | 27.68 | **+44.4%** | 12.87s → 8.21s (**−36.2%**) | 26.48ms → 19.76ms (**−25.4%**) | 26.08s → 18.07s (**−30.7%**) |
| 2 | 27.70 | 35.64 | **+28.7%** | 20.19s → 7.84s (**−61.2%**) | 30.27ms → 30.33ms (+0.2%) | 35.29s → 22.97s (**−34.9%**) |
| 4 | 36.09 | 44.65 | **+23.7%** | 20.33s → 8.10s (**−60.2%**) | 49.04ms → 46.90ms (**−4.4%**) | 44.80s → 31.50s (**−29.7%**) |
| 8 | 38.84 | 51.26 | **+32.0%** | 44.01s → 17.07s (**−61.2%**) | 80.16ms → 67.00ms (**−16.4%**) | 84.01s → 50.50s (**−39.9%**) |
| 16 | 43.87 | 53.44 | **+21.8%** | 33.51s → 29.82s (**−11.0%**) | 165.90ms → 144.45ms (**−12.9%**) | 116.29s → 101.90s (**−12.4%**) |

## What changed

Before this patch, each 8K prefill chunk could replay prior compressed KV history, expand/project it again, invoke prefix attention, and merge softmax state. The new NoPE path writes compressed KV through the normal cache-owned locations, expands the complete fitting history once, and makes one bottom-right-causal call to the existing Gluon MLA prefill kernel.

Oversized histories and RoPE MLA retain the established bounded chunk-replay path.

## Benchmark configuration

- 8× MI350X (`gfx950`), TP8/EP1
- Kimi-K3 MXFP4 target plus EAGLE3 MLA drafter
- 50,000 input tokens / 500 output tokens, `ignore_eos=true`
- Concurrency 1, 2, 4, 8, 16
- Chunked prefill 8192, MLA chunk multiplier 8
- FP8 KV cache
- Prefix caching, KVStore, mixed batching, and prefill graphs disabled
- One excluded warmup wave followed by one measured wave per concurrency
- All 31 measured requests succeeded; reported cache-hit rate was 0%

These are controlled single-wave comparisons, not multi-run confidence intervals.

## Validation

- Kimi-K3 runtime tests: 6 passed
- gfx950 Gluon MLA/KDA tests: 192 passed
- `pre-commit run --all-files`: passed

Branch: [`raikonen/optimized_k3_prefill`](https://github.com/raikonenfnu/tokenspeed/tree/raikonen/optimized_k3_prefill)

**Comment by raikonenfnu** [comment]
## Worked example: what the one-pass K3 MLA optimization actually does

This patch is primarily a **dataflow/orchestration optimization around the existing Gluon MLA prefill kernel**. It does not change the attention result, permanently store expanded K/V, or change decode.

### 1. The objects involved

For Kimi-K3 on TP8, each rank has 12 of the model's 96 attention heads. For every token, the paged MLA cache stores the compact latent representation:

- `kv_a`: 512 elements;
- `k_pe`: 64 elements;
- total cached row: 576 elements.

Before dense MLA attention, `kv_b_proj` expands each 512-element latent into the rank-local explicit values:

- 12 heads × 128 `K_nope` elements;
- 12 heads × 128 `V` elements;
- or 3,072 projected elements per token, plus the shared 64-element `k_pe` component.

Keeping the cache compressed is important for capacity. Materializing explicit K/V is temporary prefill work.

K3 has 24 MLA layers and 69 KDA linear-attention layers. This particular optimization changes the 24 MLA layers; it does not replace or optimize the KDA recurrence.

### 2. A three-token toy example

Suppose four tokens are already cached and the scheduler admits three new tokens:

```text
cached prefix P = [k0, k1, k2, k3]       (P = 4)
new suffix    E = [k4, k5, k6]           (E = 3)
queries         = [q4, q5, q6]
```

The required causal visibility is:

| Query | k0 | k1 | k2 | k3 | k4 | k5 | k6 |
|---|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
| q4 | ✓ | ✓ | ✓ | ✓ | ✓ |  |  |
| q5 | ✓ | ✓ | ✓ | ✓ | ✓ | ✓ |  |
| q6 | ✓ | ✓ | ✓ | ✓ | ✓ | ✓ | ✓ |

This is called **bottom-right causal alignment**. Q contains only the three suffix rows while K/V contains all seven rows, so the causal diagonal is shifted right by `P=4`. In general, suffix query row `r` may see key positions `j <= P + r`.

### 3. Baseline execution

The baseline computes the same table in two pieces:

1. Expand the three new latent rows into explicit K/V.
2. Run causal attention for `Q(E) × K(E)`—the triangular suffix block.
3. Read the four compressed prefix rows from the paged cache.
4. Expand the prefix latent rows with `kv_b_proj`.
5. Run noncausal attention for `Q(E) × K(P)`—the rectangular prefix block.
6. Merge the two online-softmax states `(output, log-sum-exp)` with `attn_merge_state`.

For one query, the merge is mathematically:

```text
L = log(exp(L_prefix) + exp(L_suffix))
O = exp(L_prefix - L) * O_prefix + exp(L_suffix - L) * O_suffix
```

The implementation uses the stable equivalent of this equation. It is correct, but it creates two attention invocations, requests LSE state from both, writes/reads intermediate output, and launches a merge kernel.

### 4. Optimized execution

When the history fits and the model uses NoPE MLA, the patch instead:

1. Writes the three new **compressed** latent rows into their ordinary cache-owned slots.
2. Resolves the logical seven-token history through the existing page table into `full_kv_indices`.
3. Gathers all seven compressed rows.
4. Runs one packed `kv_b_proj` over the seven rows and prepares FP8 K/V.
5. Calls the existing Gluon MLA prefill implementation once with:
   - Q length = 3;
   - K/V length = 7;
   - `causal=True` with bottom-right alignment.
6. Writes the final output directly. There is no prefix/suffix softmax merge.

In symbols, both paths compute:

```text
O_r = softmax([q_r K_prefix^T, q_r K_suffix^T] + causal_mask_r)
      @ [V_prefix; V_suffix]
```

The baseline partitions this softmax and recombines it; the optimized path presents the concatenated history to one kernel.

### 5. The actual 50K/500 benchmark shape

With `--chunked-prefill-size 8192`, a 50,000-token prompt is admitted as:

```text
8192, 8192, 8192, 8192, 8192, 8192, 848
```

On the final prefill round:

```text
P = 49,152 cached tokens
E =    848 new/query tokens
S = 50,000 total visible KV tokens
```

The baseline MLA layer performs:

```text
causal suffix attention:     Q[848,12,192] × K[848,12,192]
prefix attention:            Q[848,12,192] × K[49152,12,192]
softmax-state merge:         merge(prefix, suffix)
```

The optimized layer performs:

```text
one bottom-right causal call: Q[848,12,192] × K[50000,12,192]
```

The theoretical allowed QK pairs are the same. The benefit comes from eliminating split orchestration and from presenting a larger, regular packed problem to the Gluon kernel.

For one 50K request whose full history remains eligible:

- seven scheduler prefill rounds still occur;
- round 1 already needs one attention call in either design;
- rounds 2–7 go from two attention calls plus one merge to one attention call;
- across 24 MLA layers, that removes **144 split attention invocations and 144 softmax-state merges**.

This is why the improvement is mostly TTFT/prefill even though the attention formula and asymptotic FLOP count are unchanged.

### 6. How the code selects the path

`build_full_prefill_metadata_arrays` computes a conservative materialization guard:

```text
capacity = chunked_prefill_size × mla_chunk_multiplier
eligible when max_sequence_length <= capacity / active_prefill_batch_size
```

For the benchmark, the capacity is `8192 × 8 = 65,536` rows. Concurrency is the number of outstanding requests; it is not automatically the number of requests packed into one prefill forward. A round where one request owns the 8K prefill budget has an active prefill batch size of one, so its 50K history fits the 65,536-row capacity.

The code then follows these steps:

1. Backends attach `full_kv_indices`, `full_seq_lens`, `cu_full_seq_lens`, and `max_full_seq_len` to the ordinary prefill metadata.
2. `DeepseekV3AttentionMLA._attn` chooses full-history execution only when that metadata exists and `rotary_emb is None`.
3. `forward_normal_chunked_q_cache_prepare` prepares Q and writes the current compressed latent rows without first materializing current K/V.
4. `forward_normal_chunked_full_kv_core` gathers the full paged history, runs `kv_b_proj`, packs FP8 K/V, and invokes `forward_extend_chunked` once.

The patch also recognizes K3's `hybrid_linear_attn` wrapper as an MLA-kernel backend. That lets its full-attention sub-backend use the intended FP8 Gluon preparation path rather than losing the optimization merely because MLA is wrapped alongside KDA.

### 7. Safety boundaries and non-goals

- If the packed history exceeds capacity, `full_kv_indices` is `None` and execution falls back to the existing bounded prefix-replay path.
- RoPE MLA also retains the existing path; this fast path is intentionally NoPE-only because position-dependent K preparation has a different contract.
- The compressed paged cache remains the source of truth and is still allocated, prefix-matched, and freed by the normal cache subsystem.
- Decode still uses the existing MLA decode path.
- Scheduler chunking still exists; the patch optimizes each eligible prefill forward rather than admitting the entire 50K prompt in one scheduler step.
- This is not fewer mathematical attention interactions. It is fewer launches, fewer temporary states, no split-softmax merge, and better whole-kernel geometry.

That last distinction is the core idea: **we did not invent cheaper attention; we stopped decomposing an already-fitting attention problem into smaller pieces that were more expensive to execute on the GPU.**

**Comment by raikonenfnu** [comment]
## Implementation pseudocode: baseline vs. optimized K3 MLA prefill

The important distinction is that **scheduler chunking is unchanged**. The scheduler still admits at most `chunked_prefill_size = 8192` prompt tokens per step. This optimization changes how an MLA layer processes the cache history during one such step.

### Common scheduler loop (before and after)

```text
while any request still has prompt tokens:
    work = scheduler.admit_extend_tokens(max_total_tokens = 8192)

    for transformer_layer in model:
        hidden = layer.mla_prefill(work, hidden)
```

So this is not “disable chunked prefill.” It is a one-pass attention/materialization plan *inside each scheduled MLA layer invocation*.

Definitions used below:

```text
Q_extend       = queries for newly admitted prompt tokens
KV_extend      = compressed KV for those new tokens
KV_prefix      = compressed KV already present in the paged cache
KV_full        = KV_prefix || KV_extend
P              = number of bounded prefix-history pieces
B              = number of sequences in this scheduled extend batch
capacity       = chunked_prefill_size * mla_chunk_multiplier
               = 8192 * 8 = 65,536 tokens in the benchmark
```

`B` is the engine's current scheduled batch size, not client concurrency. For example, concurrency 16 does not imply `B == 16`; the 8192-token scheduler budget commonly admits long prefills separately.

### Baseline (`main`): bounded history replay

The clean-main dispatcher can also choose the existing absorbed MLA path. The pseudocode below expands the normal chunked path that the new full-history plan replaces:

```text
function baseline_mla_prefill(metadata, Q_raw, KV_extend_compressed):
    if metadata.use_absorbed_cached_extend:
        return absorbed_cached_extend(Q_raw, KV_extend_compressed, metadata)

    # Prepare the current scheduled chunk.
    Q = prepare_queries(Q_raw)
    K_extend, V_extend = kv_b_proj(KV_extend_compressed)
    cache.write_compressed(metadata.extend_slots, KV_extend_compressed)

    # Causal attention among the newly admitted tokens.
    output, lse = gluon_mla_attention(
        Q,
        K_extend,
        V_extend,
        q_lens = metadata.extend_seq_lens,
        kv_lens = metadata.extend_seq_lens,
        causal = true,
    )

    # Replay prior compressed history in bounded pieces.
    for prefix_piece in metadata.prefix_materialization_pieces:
        KV_piece_compressed = cache.gather(prefix_piece.slots)
        K_piece, V_piece = kv_b_proj(KV_piece_compressed)

        piece_output, piece_lse = gluon_mla_attention(
            Q,
            K_piece,
            V_piece,
            q_lens = metadata.extend_seq_lens,
            kv_lens = prefix_piece.seq_lens,
            causal = false,          # every prefix token is visible
        )

        # Exact online-softmax merge, not an approximation.
        output, lse = merge_softmax_state(
            output, lse, piece_output, piece_lse
        )

    return output
```

For `P` prefix pieces, this normal path performs `1 + P` attention launches and `P` softmax-state merges. It also invokes `kv_b_proj` separately for the current chunk and each history piece.

### Optimized K3 NoPE path: packed full history, one Gluon attention call

The metadata builder first decides whether the complete paged history can be materialized safely:

```text
function plan_full_history(seq_lens, page_tables):
    B = len(seq_lens)

    if B == 0:
        return unavailable

    if B * max(seq_lens) > capacity:
        return unavailable

    return resolve_paged_slots_for_all_prefix_and_extend_tokens(page_tables)
```

The layer then selects the fast path only for NoPE MLA when this plan exists:

```text
function optimized_mla_prefill(metadata, Q_raw, KV_extend_compressed):
    can_use_one_pass = (
        model_uses_no_rope
        and metadata.full_kv_slots is available
    )

    if not can_use_one_pass:
        return baseline_mla_prefill(
            metadata, Q_raw, KV_extend_compressed
        )

    Q = prepare_queries(Q_raw)

    # Preserve the normal cache ownership/lifetime contract.
    cache.write_compressed(metadata.extend_slots, KV_extend_compressed)

    # Gather prefix + new tokens and expand them in one packed projection.
    KV_full_compressed = cache.gather(metadata.full_kv_slots)
    K_full, V_full = kv_b_proj(KV_full_compressed)

    return gluon_mla_attention(
        Q,
        K_full,
        V_full,
        q_lens = metadata.extend_seq_lens,
        kv_lens = metadata.full_seq_lens,
        causal = true,               # bottom-right causal mask
    )
```

“Bottom-right causal” is the key correctness detail. Query row `i` belongs to the suffix (`KV_extend`), so it may see every prefix token and extend rows `0..i`, but no future extend row. Therefore the single operation is mathematically equivalent to:

```text
softmax([Q @ K_prefix^T, causal(Q @ K_extend^T)]) @ [V_prefix; V_extend]
```

The baseline computes those logits in pieces and exactly merges their softmax states; the optimized path presents the same logical logits to one attention invocation.

### Per-layer, per-scheduler-step work

| Work | Baseline normal chunked path | Optimized fitting NoPE path |
|---|---:|---:|
| Scheduler/model forwards | 1 | 1 |
| `kv_b_proj` launches | `1 + P` | `1` packed launch |
| Gluon attention launches | `1 + P` | `1` |
| Softmax-state merges | `P` | `0` |
| Compressed cache format/ownership | unchanged | unchanged |
| New attention kernel arithmetic | none | none |

This explains how performance improves without adding or modifying the Gluon attention kernel: the patch changes **packing, projection granularity, and launch topology** around the existing native gfx950 Gluon kernel. It removes repeated dispatch and merge overhead and gives the kernel one larger, denser operation.

The existing absorbed path is still the fallback when the one-pass conditions are not met. A fitting NoPE request is checked first and deliberately takes the packed full-history path; RoPE models and histories exceeding the conservative capacity keep the prior bounded behavior.

Relevant implementation:

- [NoPE fast-path selection, packed projection, and one attention call](https://github.com/raikonenfnu/tokenspeed/blob/e3185b838cbfdbce873aa07498325414ea939264/python/tokenspeed/runtime/models/deepseek_v3.py#L1095-L1237)
- [Full-history capacity check and paged-slot metadata](https://github.com/raikonenfnu/tokenspeed/blob/e3185b838cbfdbce873aa07498325414ea939264/python/tokenspeed/layers/attention/base_attn_backend.py#L203-L238)
